src/ui/public/vis/editors/default/default.js

```diff
@@ -127,6 +138,42 @@ const defaultEditor = function ($rootScope, $compile) {
             catch (e) {} // eslint-disable-line no-empty
           }, true);
 
+          // fires when visualization state changes, and we need to copy changes to editorState
+          $scope.$watch(() => {
+            return $scope.vis.getCurrentState(false);
+          }, (newState) => {
+            const updateEditorStateWithChanges = (newState, oldState, editorState) => {
+              for (const prop in newState) {
+                if (newState.hasOwnProperty(prop)) {
+                  const newStateValue = newState[prop];
+                  const oldStateValue = oldState[prop];
+                  const editorStateValue = editorState[prop];
+
+                  if (typeof newStateValue === 'object') {
+                    if (editorStateValue) {
```

Comment:
Your explanation about this sounds great. Makes sense to me. Can we add a comment?

```js
// When a node is an object we need to recurse, as we want to only update specific keys.
// But if the object is not found in the editorState, then it's safe to update the node with
// the entire object.
```